reference/best-practices/java-app.md

```diff
@@ -0,0 +1,362 @@
+
+By executing `jstack pid`, you can output the IDs and stack information of all threads in the target process. There is only the Java stack in the default output. If you want to output the C++ stack in the JVM at the same time, add the `-m` option.
+
+By using jstack multiple times, you can easily locate the stuck issue (such as updating only through Mybatis BatchExecutor flush) or the deadlock issue (such as the failure to send an SQL statement because test programs are preempting a lock in the application)
+
+In addition, `top -p $ PID -H` or Java swiss knife are common methods to view the thread ID. Also, to locate the issue of "a thread occupies a lot of CPU resources and I don't know what it is executing", do the following steps:
+
+- Use `printf "%x\n" pid` to convert the thread ID to hexadecimal
+- Go to the jstack output to find the stack information of the corresponding thread
+
+#### jmap & mat
+
+Unlike pprof/heap in Go, [jmap](https://docs.oracle.com/javase/7/docs/technotes/tools/share/jmap.html) dumps the memory snapshot of the entire process (in Go, it is the sampling of the distributor), and then the snapshot can be analyzed by another tool [mat](https://www.eclipse.org/mat/).
+
+Through mat, you can see the associated information and attributes of all objects in the process, and you can also observe the running status of the thread. For example, you can use mat to find out how many MySQL connection objects exist in the current application, and what is the address and status information of each connection object.
+
+Note that mat only handles reachable objects by default. If you want to troubleshoot young GC issues, you can adjust mat configuration to view unreachable objects. In addition, for investigating the memory allocation of young GC issues (or a large number of short-lived objects), using Java Flight Recorder is more convenient.
+
+#### trace
+
+Online applications usually do not support modifying the code, but it is often desired that dynamic instrumentation is performed in Java to locate issues. Therefore, using btrace or arthas trace is a good option. They can dynamically insert trace code without restarting the application process.
+
+#### Flame graph
+
+Obtaining flame graphs in Java applications is tedious. For details, see [Java Flame Graphs Introduction: Fire For Everyone!](http://psy-lob-saw.blogspot.com/2017/02/flamegraphs-intro-fire-for-everyone.html).
+
+## Conclusion
+
+Based on commonly used Java components that interact with databases, this document describes the common problems and solutions for developing Java applications with TiDB. TiDB is highly compatible with the MySQL protocol, so most of the best practices for MySQL-based Java applications also apply to TiDB.
+
+Join us at [ASK TUG](https://asktug.com/), and share with broad TiDB user group about the practical experience or problems you run into when you use TiDB with Java applications.
```

Comment:
maybe we can delete asktug link for English user

Reply:
Right.